File: JBTheatreToolsAndroid/app/src/main/java/com/jamesbreedon/jbtheatretools/core/SetupProfile.kt
package com.jamesbreedon.jbtheatretools.core

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

/**
 * A machine's setup — which apps are installed at which versions, which are held, and optionally the desktop
 * list layout — saved to a file and replayed on another device. The same format as the desktop launchers
 * (schema 1): `{"kind":"jbtheatretools-setup","schemaVersion":1,"createdAt","createdBy","apps":[{"id","variant"?,
 * "version","held"?}],"layout":{...}}`. Parsing is strict about shape and size.
 */
data class SetupProfile(
    val createdAt: String = "",
    val createdBy: String = "",
    val apps: List<Entry> = emptyList(),
    val layout: Layout? = null,
) {
    data class Entry(val id: String, val variant: String?, val version: String?, val held: Boolean)

    data class Layout(
        val pinned: List<String>, val hidden: List<String>, val order: List<String>,
        val categoryOrder: List<String>, val collapsed: List<String>,
    )

    class FormatError(message: String) : Exception(message)

    companion object {
        const val KIND = "jbtheatretools-setup"
        const val SCHEMA_VERSION = 1
        const val MAX_BYTES = 1_000_000
        const val MAX_ENTRIES = 500
        private val pretty = Json { prettyPrint = true }

        fun parse(json: String): SetupProfile {
            if (json.toByteArray(Charsets.UTF_8).size > MAX_BYTES) throw FormatError("This file is too large to be a setup file.")
            val root = runCatching { Json.parseToJsonElement(json) }.getOrNull() as? JsonObject
                ?: throw FormatError("This file isn't a JB Theatre Tools setup file.")
            if (str(root, "kind") != KIND) throw FormatError("This file isn't a JB Theatre Tools setup file.")
            val schema = (root["schemaVersion"] as? JsonPrimitive)?.takeIf { !it.isString }?.intOrNull
            if (schema == null || schema < 1) throw FormatError("This setup file has no valid schemaVersion.")
            if (schema > SCHEMA_VERSION) throw FormatError("This setup file was made by a newer JB Theatre Tools — update the launcher first.")
            val apps = root["apps"] as? JsonArray ?: throw FormatError("This setup file lists no apps.")
            val entries = ArrayList<Entry>()
            for (a in apps) {
                if (entries.size >= MAX_ENTRIES) break
                val o = a as? JsonObject ?: continue
                val id = str(o, "id")?.trim()
                if (id.isNullOrEmpty()) continue
                val variant = str(o, "variant")?.trim()
                val version = str(o, "version")?.trim()
                val held = (o["held"] as? JsonPrimitive)?.takeIf { !it.isString }?.booleanOrNull == true
                entries.add(Entry(id, variant?.ifEmpty { null }, version?.ifEmpty { null }, held))
            }
            val l = root["layout"] as? JsonObject
            val layout = l?.let {
                Layout(ids(it, "pinned"), ids(it, "hidden"), ids(it, "order"), ids(it, "categoryOrder"), ids(it, "collapsed"))
            }
            return SetupProfile(str(root, "createdAt") ?: "", str(root, "createdBy") ?: "", entries, layout)
        }

        private fun str(o: JsonObject, name: String): String? =
            (o[name] as? JsonPrimitive)?.takeIf { it.isString }?.content

        private fun ids(o: JsonObject, name: String): List<String> {
            val arr = o[name] as? JsonArray ?: return emptyList()
            return arr.mapNotNull { (it as? JsonPrimitive)?.takeIf { p -> p.isString }?.content }
                .filter { it.isNotBlank() }.take(MAX_ENTRIES)
        }

        private val utc: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'").withZone(ZoneOffset.UTC)

        fun timestamp(now: Instant): String = utc.format(now)

        /** A suggested file name: "JB Theatre Tools setup 2026-09-25.json". */
        fun suggestedFileName(today: LocalDate): String = "JB Theatre Tools setup $today.json"
    }

    fun serialize(): String = pretty.encodeToString(JsonElement.serializer(), buildJsonObject {
        put("kind", KIND)
        put("schemaVersion", SCHEMA_VERSION)
        put("createdAt", createdAt)
        put("createdBy", createdBy)
        put("apps", buildJsonArray {
            for (e in apps) add(buildJsonObject {
                put("id", e.id)
                e.variant?.let { put("variant", it) }
                e.version?.let { put("version", it) }
                if (e.held) put("held", true)
            })
        })
        layout?.let { l ->
            put("layout", buildJsonObject {
                putJsonArray("pinned") { l.pinned.forEach { add(JsonPrimitive(it)) } }
                putJsonArray("hidden") { l.hidden.forEach { add(JsonPrimitive(it)) } }
                putJsonArray("order") { l.order.forEach { add(JsonPrimitive(it)) } }
                putJsonArray("categoryOrder") { l.categoryOrder.forEach { add(JsonPrimitive(it)) } }
                putJsonArray("collapsed") { l.collapsed.forEach { add(JsonPrimitive(it)) } }
            })
        }
    })
}

/** What importing a setup file would do on this device — computed before anything installs (the preview). */
object SetupPlanner {
    data class CatalogEntry(val id: String, val name: String, val variants: List<Pair<String, String>>)

    /** One slot to install. [variantId] null = the default edition; [tag] null = the latest release. */
    data class Install(val appId: String, val variantId: String?, val tag: String?, val label: String)

    data class Plan(
        val toInstall: List<Install>, val alreadyInstalled: List<String>,
        val skipped: List<String>, val holdIds: List<String>,
    )

    fun build(
        profile: SetupProfile,
        catalog: List<CatalogEntry>,
        installedKeys: Set<String>,
        supportsVariants: Boolean = true,
    ): Plan {
        val byId = catalog.associateBy { it.id }
        val toInstall = ArrayList<Install>()
        val already = ArrayList<String>()
        val skipped = ArrayList<String>()
        val hold = ArrayList<String>()
        val seenKeys = HashSet<String>()
        for (e in profile.apps) {
            val app = byId[e.id]
            if (app == null) { skipped.add("${e.id} — not in this launcher's catalog"); continue }
            var variant = e.variant
            var variantLabel: String? = null
            if (variant != null) {
                val idx = app.variants.indexOfFirst { it.first == variant }
                if (idx < 0) { skipped.add("${app.name} ($variant) — no such edition"); continue }
                if (idx == 0) variant = null
                else if (!supportsVariants) { skipped.add("${app.name} (${app.variants[idx].second}) — not available here"); continue }
                else variantLabel = app.variants[idx].second
            }
            val key = if (variant == null) app.id else "${app.id}@$variant"
            if (!seenKeys.add(key)) continue
            val label = if (variantLabel == null) app.name else "${app.name} ($variantLabel)"
            if (e.held && app.id !in hold) hold.add(app.id)
            if (key in installedKeys) { already.add(label); continue }
            toInstall.add(Install(app.id, variant, if (e.held) e.version else null, label))
        }
        return Plan(toInstall, already, skipped, hold)
    }

    /** The preview text shown before an import runs. */
    fun summary(plan: Plan): String {
        val sb = StringBuilder()
        if (plan.toInstall.isEmpty()) sb.append("Nothing to install — this machine already has every app in the file.")
        else {
            sb.append("Install ${plan.toInstall.size} app${if (plan.toInstall.size == 1) "" else "s"}:")
            for (i in plan.toInstall)
                sb.append("\n  • ").append(i.label).append(if (i.tag != null) " ${VersionCompare.display(i.tag)} (held)" else "")
        }
        if (plan.alreadyInstalled.isNotEmpty())
            sb.append("\n\nAlready installed (left as they are): ${plan.alreadyInstalled.joinToString(", ")}")
        if (plan.holdIds.isNotEmpty())
            sb.append("\n\nHeld at their versions: ${plan.holdIds.size} app${if (plan.holdIds.size == 1) "" else "s"}")
        if (plan.skipped.isNotEmpty())
            sb.append("\n\nSkipped:\n  • ").append(plan.skipped.joinToString("\n  • "))
        return sb.toString()
    }
}

File: JBTheatreToolsWin/Core/SetupProfile.cs
using System.Globalization;
using System.Text;
using System.Text.Json;

namespace JBTheatreTools;

/// <summary>A machine's setup — which apps (and editions) are installed at which versions, which are held, and
/// optionally the list layout — saved to a file and replayed on another machine (any launcher: the ids are the
/// catalog's). Format (schema 1):
/// <code>{"kind":"jbtheatretools-setup","schemaVersion":1,"createdAt":"…","createdBy":"…",
///  "apps":[{"id":"nditools","variant":"full","version":"v2.0.0","held":true}],
///  "layout":{"pinned":[],"hidden":[],"order":[],"categoryOrder":[],"collapsed":[]}}</code>
/// "variant" is omitted for an app's default edition. Parsing is strict about shape and size so a stray file
/// can't do anything surprising.</summary>
public sealed class SetupProfile
{
    public const string Kind = "jbtheatretools-setup";
    public const int SchemaVersion = 1;
    public const int MaxBytes = 1_000_000;
    public const int MaxEntries = 500;

    public sealed record Entry(string Id, string? Variant, string? Version, bool Held);

    public sealed record Layout(List<string> Pinned, List<string> Hidden, List<string> Order,
                                List<string> CategoryOrder, List<string> Collapsed);

    public string CreatedAt { get; init; } = "";
    public string CreatedBy { get; init; } = "";
    public List<Entry> Apps { get; init; } = new();
    public Layout? AppLayout { get; init; }

    public static SetupProfile Parse(string json)
    {
        if (Encoding.UTF8.GetByteCount(json) > MaxBytes) throw new FormatException("This file is too large to be a setup file.");
        JsonDocument doc;
        try { doc = JsonDocument.Parse(json); }
        catch (JsonException) { throw new FormatException("This file isn't a JB Theatre Tools setup file."); }
        using (doc)
        {
            var root = doc.RootElement;
            if (root.ValueKind != JsonValueKind.Object || Str(root, "kind") != Kind)
                throw new FormatException("This file isn't a JB Theatre Tools setup file.");
            if (!root.TryGetProperty("schemaVersion", out var sv) || sv.ValueKind != JsonValueKind.Number ||
                !sv.TryGetInt32(out var schema) || schema < 1)
                throw new FormatException("This setup file has no valid schemaVersion.");
            if (schema > SchemaVersion)
                throw new FormatException("This setup file was made by a newer JB Theatre Tools — update the launcher first.");
            if (!root.TryGetProperty("apps", out var apps) || apps.ValueKind != JsonValueKind.Array)
                throw new FormatException("This setup file lists no apps.");
            var entries = new List<Entry>();
            foreach (var a in apps.EnumerateArray())
            {
                if (entries.Count >= MaxEntries) break;
                if (a.ValueKind != JsonValueKind.Object) continue;
                var id = Str(a, "id")?.Trim();
                if (string.IsNullOrEmpty(id)) continue;
                var variant = Str(a, "variant")?.Trim();
                var version = Str(a, "version")?.Trim();
                bool held = a.TryGetProperty("held", out var h) && h.ValueKind == JsonValueKind.True;
                entries.Add(new Entry(id, string.IsNullOrEmpty(variant) ? null : variant,
                                      string.IsNullOrEmpty(version) ? null : version, held));
            }
            Layout? layout = null;
            if (root.TryGetProperty("layout", out var l) && l.ValueKind == JsonValueKind.Object)
                layout = new Layout(Ids(l, "pinned"), Ids(l, "hidden"), Ids(l, "order"), Ids(l, "categoryOrder"), Ids(l, "collapsed"));
            return new SetupProfile
            {
                CreatedAt = Str(root, "createdAt") ?? "",
                CreatedBy = Str(root, "createdBy") ?? "",
                Apps = entries,
                AppLayout = layout,
            };
        }
    }

    private static string? Str(JsonElement e, string name) =>
        e.TryGetProperty(name, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() : null;

    private static List<string> Ids(JsonElement e, string name)
    {
        var list = new List<string>();
        if (!e.TryGetProperty(name, out var arr) || arr.ValueKind != JsonValueKind.Array) return list;
        foreach (var x in arr.EnumerateArray())
            if (x.ValueKind == JsonValueKind.String && !string.IsNullOrWhiteSpace(x.GetString()) && list.Count < MaxEntries)
                list.Add(x.GetString()!);
        return list;
    }

    public string Serialize()
    {
        using var ms = new MemoryStream();
        using (var w = new Utf8JsonWriter(ms, new JsonWriterOptions { Indented = true }))
        {
            w.WriteStartObject();
            w.WriteString("kind", Kind);
            w.WriteNumber("schemaVersion", SchemaVersion);
            w.WriteString("createdAt", CreatedAt);
            w.WriteString("createdBy", CreatedBy);
            w.WriteStartArray("apps");
            foreach (var e in Apps)
            {
                w.WriteStartObject();
                w.WriteString("id", e.Id);
                if (e.Variant != null) w.WriteString("variant", e.Variant);
                if (e.Version != null) w.WriteString("version", e.Version);
                if (e.Held) w.WriteBoolean("held", true);
                w.WriteEndObject();
            }
            w.WriteEndArray();
            if (AppLayout != null)
            {
                w.WriteStartObject("layout");
                void Arr(string n, List<string> v) { w.WriteStartArray(n); foreach (var s in v) w.WriteStringValue(s); w.WriteEndArray(); }
                Arr("pinned", AppLayout.Pinned);
                Arr("hidden", AppLayout.Hidden);
                Arr("order", AppLayout.Order);
                Arr("categoryOrder", AppLayout.CategoryOrder);
                Arr("collapsed", AppLayout.Collapsed);
                w.WriteEndObject();
            }
            w.WriteEndObject();
        }
        return Encoding.UTF8.GetString(ms.ToArray());
    }

    public static string Timestamp(DateTimeOffset now) =>
        now.ToUniversalTime().ToString("yyyy-MM-dd'T'HH:mm:ss'Z'", CultureInfo.InvariantCulture);

    /// <summary>A suggested file name: "JB Theatre Tools setup 2026-09-25.json".</summary>
    public static string SuggestedFileName(DateTime nowLocal) =>
        $"JB Theatre Tools setup {nowLocal.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture)}.json";
}

/// <summary>What importing a setup file would do on this machine — computed before anything installs so the user
/// sees a preview.</summary>
public static class SetupPlanner
{
    /// <summary>One catalog app as the planner needs it: its editions in order (empty = no editions).</summary>
    public sealed record CatalogEntry(string Id, string Name, IReadOnlyList<(string Id, string Label)> Variants);

    /// <summary>One slot to install. <paramref name="VariantId"/> null = the default edition;
    /// <paramref name="Tag"/> null = the latest release.</summary>
    public sealed record Install(string AppId, string? VariantId, string? Tag, string Label);

    public sealed record Plan(List<Install> ToInstall, List<string> AlreadyInstalled, List<string> Skipped, List<string> HoldIds,
        List<string>? HoldNames = null);

    /// <param name="installedKeys">Install keys present on this machine ("id" or "id@variant").</param>
    /// <param name="supportsVariants">False on Android: editions other than the default are skipped.</param>
    /// <param name="allowDevTags">False when Development builds are off here: an entry held at a development build
    /// is skipped — listed under Skipped, not installed and not held — so the preview never promises what the
    /// import won't do.</param>
    public static Plan Build(SetupProfile profile, IReadOnlyList<CatalogEntry> catalog, ISet<string> installedKeys,
                             bool supportsVariants = true, bool allowDevTags = true)
    {
        var byId = catalog.ToDictionary(c => c.Id, StringComparer.Ordinal);
        var toInstall = new List<Install>();
        var already = new List<string>();
        var skipped = new List<string>();
        var hold = new List<string>();
        var seenKeys = new HashSet<string>(StringComparer.Ordinal);
        foreach (var e in profile.Apps)
        {
            if (!byId.TryGetValue(e.Id, out var app)) { skipped.Add($"{e.Id} — not in this launcher's catalog"); continue; }
            string? variant = e.Variant;
            string? variantLabel = null;
            if (variant != null)
            {
                int idx = -1;
                for (int i = 0; i < app.Variants.Count; i++) if (app.Variants[i].Id == variant) { idx = i; break; }
                if (idx < 0) { skipped.Add($"{app.Name} ({variant}) — no such edition"); continue; }
                if (idx == 0) variant = null;                          // the default edition's own id
                else if (!supportsVariants) { skipped.Add($"{app.Name} ({app.Variants[idx].Label}) — not available here"); continue; }
                else variantLabel = app.Variants[idx].Label;
            }
            var key = variant == null ? app.Id : $"{app.Id}@{variant}";
            if (!seenKeys.Add(key)) continue;                           // listed twice
            var label = variantLabel == null ? app.Name : $"{app.Name} ({variantLabel})";
            if (e.Held && e.Version != null && !allowDevTags && VersionCompare.IsDev(e.Version))
            {
                skipped.Add($"{label} {VersionCompare.Display(e.Version)} — a development build (not switched on here)");
                continue;
            }
            if (e.Held && !hold.Contains(app.Id)) hold.Add(app.Id);
            if (installedKeys.Contains(key)) { already.Add(label); continue; }
            toInstall.Add(new Install(app.Id, variant, e.Held ? e.Version : null, label));
        }
        return new Plan(toInstall, already, skipped, hold, hold.Select(id => byId.TryGetValue(id, out var c) ? c.Name : id).ToList());
    }

    /// <summary>The preview text shown before an import runs.</summary>
    public static string Summary(Plan plan)
    {
        var sb = new StringBuilder();
        if (plan.ToInstall.Count == 0) sb.Append("Nothing to install — this machine already has every app in the file.");
        else
        {
            sb.Append($"Install {plan.ToInstall.Count} app{(plan.ToInstall.Count == 1 ? "" : "s")}:");
            foreach (var i in plan.ToInstall)
                sb.Append("\n  • ").Append(i.Label).Append(i.Tag != null ? $" {VersionCompare.Display(i.Tag)} (held)" : "");
        }
        if (plan.AlreadyInstalled.Count > 0)
            sb.Append($"\n\nAlready installed (left as they are): {string.Join(", ", plan.AlreadyInstalled)}");
        if (plan.HoldIds.Count > 0)
            sb.Append($"\n\nHeld at their versions: {plan.HoldIds.Count} app{(plan.HoldIds.Count == 1 ? "" : "s")}"
                + (plan.HoldNames is { Count: > 0 } n ? $" ({string.Join(", ", n)})" : "")
                + " — Update All and automatic updates leave them at the version this machine has");
        if (plan.Skipped.Count > 0)
            sb.Append("\n\nSkipped:\n  • ").Append(string.Join("\n  • ", plan.Skipped));
        return sb.ToString();
    }
}
